src/service/resources-impl.js

```diff
@@ -239,43 +278,147 @@ export class ResourcesImpl {
     }
   }
 
+  /**
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   * @param {!IntersectionObserver} unusedObserver
+   * @private
+   */
+  intersects_(entries, unusedObserver) {
+    dev().fine(TAG_, 'intersect', entries);
+
+    const toUnload = [];
+
+    const promises = entries.map(entry => {
+      const {boundingClientRect, isIntersecting, target, rootBounds} = entry;
+      // Closure Compiler doesn't recognize boundingClientRect as a ClientRect.
+      const clientRect = /** @type {!ClientRect} */ (boundingClientRect);
+      devAssert(target.isUpgraded());
+      const r = devAssert(Resource.forElementOptional(target));
+
+      // discoverWork_():
+      // [x] Phase 1: Build and relayout as needed. All mutations happen here.
+      // [x]   1A: Apply sizes/media-queries to un-measured/un-laid-out resources.
+      // [~] Phase 2: Remeasure if there were any relayouts. All reads happen here.
+      // [x]   2A: Unload non-displayed resources.
+      // [x] Phase 3: Trigger "viewport enter/exit" events.
+      // [x] Phase 4: Schedule elements for layout within a reasonable distance from current viewport.
+      // [x]   4A: Force build for all resources visible, measured, and in the viewport.
+      // [ ] Phase 5: Idle Render Outside Viewport layout: layout up to 4 items with idleRenderOutsideViewport true.
+      // [ ] Phase 6: Idle layout: layout more if we are otherwise not doing much.
+
+      // Force all intersecting, non-zero-sized, non-owned elements to be built.
+      // E.g. ensures that all in-viewport elements are built in prerender mode.
+      if (
+        !r.isBuilt() &&
+        !r.isBuilding() &&
+        isIntersecting &&
+        r.isDisplayed(clientRect) &&
+        !r.hasOwner()
+      ) {
+        // TODO(willchou): Can this cause scroll jank since we no longer wait
+        // for scrolling to stop?
+        this.buildOrScheduleBuildForResource_(
+          r,
+          /* checkForDupes */ true,
+          /* scheduleWhenBuilt */ false,
+          /* force */ true
+        );
+        dev().fine(TAG_, 'force build:', r.debugid);
+      }
+
+      return r.whenBuilt().then(() => {
+        const wasIntersecting = r.isInViewport();
+        let noLongerDisplayed = this.measureResource_(r, clientRect);
+
+        // Sometimes the intersection callback is too early to recognize
+        // client rect changes due to animations (e.g. amp-accordion).
+        // These cases can still be detected since isIntersecting == false
+        // despite the target element overlapping the root's bounds.
+        if (
+          !noLongerDisplayed &&
+          wasIntersecting &&
+          !isIntersecting &&
+          layoutRectsOverlap(clientRect, rootBounds)
+        ) {
+          // TODO(willchou): Sometimes causes an unnecessary unload when
```

Comment:
Seems like a major issue? Also, what kind of animation we're talking about here? This would be an issue only when we animate `display:none`, but we never do that. Could you clarify?